tests/SAML2/CustomBaseID.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace SAML2;
+
+use DOMElement;
+use SAML2\XML\saml\AbstractBaseIDType;
+use SAML2\XML\saml\BaseID;
+use Webmozart\Assert\Assert;
+
+class CustomBaseID extends AbstractBaseIDType
+{
+    /** @var float $value */
+    private $value;
+
+
+    public function __construct(float $value, string $NameQualifier = null, string $SPNameQualifier = null)
```

Comment:
I'm tempted to say that if a custom ID can have a value with a different data type, then the class implementing that ID should convert that into string first, then pass it up to `BaseID::__construct()`.

Reply:
Maybe it can't, because the type should point to a class..  I did add a getvalue/setvalue to the base class

Reply:
I'm not entirely sure I follow. The thing is, as far as I can tell, the value can be anything. But if the value has any type, we can't generate the XML from it. So the only option then is to accept he string representation of the value from the beginning.

Reply:
Yes, but the type has to be a class right now (like CustomBaseID) and is being checked against the class registry (which would fail for type="xs:float")

Reply:
That's the thing, `xsi:type` could be anything. We shouldn't enforce class names as values of `xsi:type`, because `xsi:type` is a [`QName`](https://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/datatypes.html#QName) and as such, it can contain characters that are not suitable for class names in PHP. Imagine there is a given type of ID that becomes popular, and its `xsi:type` is defined as `urn:whatever:id`. Enforcing the name of the extending class to match `xsi:type` means effectively it would be impossible to implement such an ID in our library.